EIP4494: Permits for ERC721 (#4494)
ethereum/EIPs: EIPS/eip-4494.md

---
eip: 4494

> title: Permit for ERC721 NFTs

Reviewer: ```suggestion
title: Permit for ERC-721 NFTs
```

> description: ERC712-singed approvals for ERC721 NFTs

Reviewer: ```suggestion
description: ERC-712 signed approvals for ERC-721 NFTs
```

author: Simon Fremaux (@dievardump), William Schwab (@wschwab)
discussions-to: https://ethereum-magicians.org/t/eip-extending-erc2612-style-permits-to-erc721-nfts/7519/2
status: Draft
type: Standards Track
category: ERC
created: 2021-11-25
requires: 165, 712, 721
---

## Abstract

> The "Permit" approval flow outlined in [EIP-2612](./eip-2612.md) has proven a very valuable advancement in UX by creating gasless approvals for ERC20 tokens. This EIP extends the pattern to ERC721 NFTs. This EIP borrows heavily from EIP-2612.

Reviewer: ```suggestion
The "Permit" approval flow outlined in [EIP-2612](./eip-2612.md) has proven to be a very valuable advancement in UX by creating gasless approvals for [ERC-20](./eip-20.md) tokens. This EIP extends the pattern to [ERC-721](./eip-721) NFTs. This EIP borrows heavily from [EIP-2612](./eip-2612).
```

> The "Permit" approval flow outlined in [ERC-2612](./eip-2612.md) has proven a very valuable advancement in UX by creating gasless approvals for ERC20 tokens. This EIP extends the pattern to ERC-721 NFTs. This EIP borrows heavily from ERC-2612.

Reviewer: ```suggestion
The "Permit" approval flow outlined in [ERC-2612](./eip-2612.md) has proven a very valuable advancement in UX by creating gasless approvals for ERC-20 tokens. This EIP extends the pattern to ERC-721 NFTs. This EIP borrows heavily from ERC-2612.
```

Author: ooh, I missed one, I'll get it in the next push

> This requires a separate EIP due to the difference in structure between ERC20 and ERC721 tokens. While ERC20 permits use value (the amount of the ERC20 token being approved) and a nonce based on the owner's address, ERC721 permits focus on the `tokenId` of the NFT and increment nonce based on the transfers of the NFT.

Reviewer: ```suggestion
This requires a separate EIP due to the difference in structure between ERC-20 and ERC-721 tokens. While ERC-20 permits use value (the amount of the ERC-20 token being approved) and a nonce based on the owner's address, ERC-721 permits focus on the `tokenId` of the NFT and increment nonce based on the transfers of the NFT.
```

## Motivation

> The permit structure outlined in [EIP-2612](./eip-2612.md) allows for a signed message (structured as outlined in [EIP-712](./eip-712.md)) to be used in order to create an approval. Whereas the normal approval-based pull flow generally involves two transactions, one to approve a contract and a second for the contract to pull the asset, which is poor UX and often confuses new users, a permit-style flow only requires signing a message and a transaction. Additional information can be found in [EIP-2612](./eip-2612.md).

Reviewer: ```suggestion
The permit structure outlined in [EIP-2612](./eip-2612.md) allows for a signed message (structured as outlined in [ERC-712](./eip-712.md)) to be used in order to create an approval. Whereas the normal approval-based pull flow generally involves two transactions, one to approve a contract and a second for the contract to pull the asset, which is poor UX and often confuses new users, a permit-style flow only requires signing a message and a transaction. Additional information can be found in [EIP-2612](./eip-2612.md).
```

Author: When should standards be referred to as an EIP and when as an ERC?

(I went for colloquial references being ERC, but actual references to standards being EIP)

> [EIP-2612](./eip-2612.md) only outlines a permit architecture for ERC20 tokens. This ERC proposes an architecture for ERC721 NFTs, which also contain an approve architecture that would benefit from a signed message-based approval flow.

Reviewer: ```suggestion
[EIP-2612](./eip-2612.md) only outlines a permit architecture for ERC-20 tokens. This ERC proposes an architecture for ERC-721 NFTs, which also contain an approve architecture that would benefit from a signed message-based approval flow.
```

## Specification

> The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Reviewer: ```suggestion
```
Now implied per EIP-1.

Reviewer: I'm wrong - this is correct as-is.

> Three new functions MUST be added to [ERC721](./eip-721.md):

Reviewer: ```suggestion
Three new functions MUST be added to [ERC-721](./eip-721.md):
```

```solidity
pragma solidity 0.8.10;

import "./IERC165.sol";

///
/// @dev Interface for token permits for ERC721
///
interface IERC4494 is IERC165 {
  /// ERC165 bytes to add to interface array - set in parent contract
  ///
  /// _INTERFACE_ID_ERC4494 = 0x5604e225

  /// @notice Function to approve by way of owner signature
  /// @param spender the address to approve
  /// @param tokenId the index of the NFT to approve the spender on
  /// @param deadline a timestamp expiry for the permit
  /// @param sig a traditional or EIP2098 signature
  function permit(address spender, uint256 tokenId, uint256 deadline, bytes memory sig) external;
  /// @notice Returns the nonce of an NFT - useful for creating permits
  /// @param tokenId the index of the NFT to get the nonce of
  /// @return the uint256 representation of the nonce
  function nonces(uint256 tokenId) external view returns(uint256);
  /// @notice Returns the domain separator used in the encoding of the signature for permits, as defined by EIP712
  /// @returns the bytes32 domain separator
  function DOMAIN_SEPARATOR() external view returns(bytes32);
}
```
The semantics of which are as follows:

> For all addresses `spender`, `uint256`s `tokenId, `deadline`, and `nonce`, and `bytes` `sig`, a call to `permit(spender, tokenId, deadline, sig)` MUST set `spender` as approved on `tokenId` as long as the owner of `tokenId` remains in possession of it, and MUST emit a corresponding `Approval` event, if and only if the following conditions are met:

Reviewer 2: missing a ` after the tokenId ?

Author: yup, good catch

## Rationale
The `permit` function is sufficient for enabling a `safeTransferFrom` transaction to be made without the need for an additional transaction.

The format avoids any calls to unknown code.

The `nonces` mapping is given for replay protection.

A common use case of permit has a relayer submit a Permit on behalf of the owner. In this scenario, the relaying party is essentially given a free option to submit or withhold the Permit. If this is a cause of concern, the owner can limit the time a Permit is valid for by setting deadline to a value in the near future. The deadline argument can be set to uint(-1) to create Permits that effectively never expire.

ERC-712 typed messages are included because of its use in [EIP-2612](./eip-2612.md), which in turn cites widespread adoption in many wallet providers.

> While EIP-2612 focuses on the value being approved, this EIP focuses on the `tokenId` of the NFT being approved via `permit`. This enables a flexibility that cannot be achieved with ERC20 (or even ERC1155) tokens, enabling a single owner to give multiple permits on the same NFT. This is possible since each ERC721 token is discrete (oftentimes referred to as non-fungible), which allows assertion that this token is still in the possession of the `owner` simply and conclusively.

Reviewer: ```suggestion
While EIP-2612 focuses on the value being approved, this EIP focuses on the `tokenId` of the NFT being approved via `permit`. This enables a flexibility that cannot be achieved with ERC-20 (or even [ERC-1155](./eip-1155.md)) tokens, enabling a single owner to give multiple permits on the same NFT. This is possible since each ERC-721 token is discrete (oftentimes referred to as non-fungible), which allows assertion that this token is still in the possession of the `owner` simply and conclusively.
```

> Whereas EIP-2612 splits signatures into their `v,r,s` components, this EIP opts to instead take a `bytes` array of variable length in order to support [EIP_2098](./eip-2098) signatures (64 bytes), which cannot be easily separated or reconstructed from `r,s,v` components (65 bytes).

Reviewer: ```suggestion
Whereas EIP-2612 splits signatures into their `v,r,s` components, this EIP opts to instead take a `bytes` array of variable length in order to support [EIP-2098](./eip-2098) signatures (64 bytes), which cannot be easily separated or reconstructed from `r,s,v` components (65 bytes).
```

## Backwards Compatibility

> There are already some ERC721 contracts implementing a `permit`-style architecture, most notably Uniswap v3. 

Reviewer: ```suggestion
There are already some ERC-721 contracts implementing a `permit`-style architecture, most notably Uniswap v3. 
```
